filecoin/activeretrievals.go

```diff
@@ -0,0 +1,252 @@
+// query candidates. It's used to increment the number of finished queries and
+// also check whether cleanup may be necessary.
+func (arm *ActiveRetrievalsManager) QueryCandidatedFinished(retrievalCid cid.Cid) {
+	arm.lk.Lock()
+	defer arm.lk.Unlock()
+
+	ar := arm.findActiveRetrievalFor(retrievalCid)
+	if ar == nil {
+		log.Errorf("Unexpected active retrieval QueryCandidatedFinished for %s (%d active, %d/%d query candidates, %d/%d retrieval candidates)", retrievalCid, len(arm.arMap), ar.queriesFinished, ar.queryCandidateCount, ar.retrievalsFinished, ar.retrievalCandidateCount)
+		return
+	}
+	ar.queriesFinished++
+	log.Debugf("QueryCandidatedFinished for %s (%d active, %d/%d query candidates, %d/%d retrieval candidates)", retrievalCid, len(arm.arMap), ar.queriesFinished, ar.queryCandidateCount, ar.retrievalsFinished, ar.retrievalCandidateCount)
+	arm.maybeFinish(retrievalCid, ar)
+}
+
+// RetrievalCandidatedFinished registers that a retrieval has finished with one
+// of the query candidates. In the case of a failure, it's used to increment the
+// number of finished so we can check the finished count against the number of
+// expected candidates for possible clean-up. In the case of success, we can
+// assume no more retrievals will occur so we can jump straight to clean-up.
+func (arm *ActiveRetrievalsManager) RetrievalCandidatedFinished(retrievalCid cid.Cid, success bool) {
+	arm.lk.Lock()
+	defer arm.lk.Unlock()
+
+	ar := arm.findActiveRetrievalFor(retrievalCid)
+	if ar == nil {
+		log.Errorf("Unexpected active retrieval RetrievalCandidatedFinished for %s (%d active, %d/%d query candidates, %d/%d retrieval candidates)", retrievalCid, len(arm.arMap), ar.queriesFinished, ar.queryCandidateCount, ar.retrievalsFinished, ar.retrievalCandidateCount)
+		return
+	}
+	ar.retrievalsFinished++
+	if success {
+		// if success, there won't be any more retrieval attempts so we can short-circuit clean-up
+		ar.retrievalCandidateCount = ar.retrievalsFinished
+	}
+	log.Debugf("RetrievalCandidatedFinished for %s (%d active, %d/%d query candidates, %d/%d retrieval candidates)", retrievalCid, len(arm.arMap), ar.queriesFinished, ar.queryCandidateCount, ar.retrievalsFinished, ar.retrievalCandidateCount)
+	arm.maybeFinish(retrievalCid, ar)
+}
+
+// findActiveRetrievalFor looks up a retrieval in the map by CID but will also
+// search for rootCid matches as a secondary option since they can also be
+// identified by that way.
+// Note that one implication of looking up a CID in both places is that calls to
+// New() will return ErrRetrievalAlreadyRunning if that CID is involved as
+// either a primary CID or a root CID. But this is reasonable since a successful
+// retrieval will yield the newly requested CID regardless.
+func (arm *ActiveRetrievalsManager) findActiveRetrievalFor(cid cid.Cid) *activeRetrieval {
+	ar, ok := arm.arMap[cid]
+	if !ok {
+		for _, ar := range arm.arMap {
+			if cid.Equals(ar.rootCid) {
+				return ar
+			}
+		}
+	}
+	return ar
```

Comment:
so ar here is nil no? I feel like we should have it say `return nil` -- moreover, if I understand this func correctly, return nil means "no record found" and I think we may want to make that explicit.

Reply:
changed to add a bool as an additional return value so it's explicit